tools/tpm2_policyauthorize.c

```diff
@@ -0,0 +1,180 @@
+   const char *to_authorize_policy_qualifier_file;
+   const char *verifying_public_key_name_file;
+   const char *verification_ticket_file;
+   const char *policy_out_path;
+};
+
+static tpm2_startauthsession_ctx ctx;
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'o':
+        ctx.policy_out_path = value;
+        break;
+    case 'S':
+        ctx.session_path = value;
+        break;
+    case 'f':
+        ctx.to_authorize_policy_file = value;
+        break;
+    case 'q':
+        ctx.to_authorize_policy_qualifier_file = value;
+        break;
+    case 'n':
+        ctx.verifying_public_key_name_file = value;
+        break;
+    case 't':
+        ctx.verification_ticket_file = value;
+        break;
+    }
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "policy-file",       required_argument, NULL, 'o' },
+        { "session",           required_argument, NULL, 'S' },
+        { "input-policy-file", required_argument, NULL, 'f' },
+        { "qualifier",         required_argument, NULL, 'q' },
+        { "name",              required_argument, NULL, 'n' },
+        { "ticket",            required_argument, NULL, 't' },
+    };
+
+    *opts = tpm2_options_new("o:S:f:q:n:t:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, 0);
+
+    return *opts != NULL;
+}
+
+bool is_check_input_options_ok(void) {
+
+    if (!ctx.session_path) {
+        LOG_ERR("Must specify -S session file.");
+        return false;
+    }
+
+    if (!ctx.to_authorize_policy_file) {
+        LOG_ERR("Must specify -a to_authorize_policy_file name.");
+        return false;
+    }
+
+    if (!ctx.verifying_public_key_name_file) {
+        LOG_ERR("Must specify -c verifying_public_key_name_file name.");
+        return false;
+    }
+
+    return true;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    if (!is_check_input_options_ok()) {
+        return -1;
+    }
+
+    int rc = 1;
+
+    tpm2_session *s = tpm2_session_restore(sapi_context, ctx.session_path);
+    if (!s) {
+        return rc;
+    }
+
+    bool result = tpm2_policy_build_policyauthorize(sapi_context, s,
+        ctx.to_authorize_policy_file, ctx.to_authorize_policy_qualifier_file,
+        ctx.verifying_public_key_name_file,ctx.verification_ticket_file);
+    if (!result) {
+        LOG_ERR("Could not build tpm authorized policy");
+        goto out;
+    }
+
+    TPM2B_DIGEST policy_digest = TPM2B_EMPTY_INIT;
+    result = tpm2_policy_get_digest(sapi_context, s, &policy_digest);
+    if (!result) {
+        LOG_ERR("Could not build tpm policy");
+        goto out;
+    }
+
+    tpm2_tool_output("policy-digest: 0x");
+    UINT16 i;
+    for(i = 0; i < policy_digest.size; i++) {
+        tpm2_tool_output("%02X", policy_digest.buffer[i]);
+    }
+    tpm2_tool_output("\n");
```

Comment:
Theirs a routine for dumping hex to output. I used in loadexternal for dumping the name, copy that.